File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

public class Solution {

    public int DiameterOfBinaryTree(TreeNode root) {
        int[] diameter = new int[1] {0};
        TraverseTree(root, diameter);
        return diameter[0];
    }

    private int TraverseTree(TreeNode root, int[] diameter) {
        if (root == null) return 0;

        int left = TraverseTree(root.left, diameter);
        int right = TraverseTree(root.right, diameter);
        diameter[0] = Math.Max(left + right, diameter[0]);
        
        return Math.Max(left, right) + 1;
    }
}

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isBalanced(TreeNode* root) {
        int result = dfs(root);
        if (result == -1) return false;
        return true;
    }

    int dfs(TreeNode* root) {
        if (root == nullptr) return 0;

        int left = dfs(root->left);
        if (left == -1) return -1;
        int right = dfs(root->right);
        if (right == -1) return -1;

        if (std::abs(left - right) > 1) return -1;
        return std::max(left, right) + 1;
    }
};
